File: simulation/solver.py
"""
Simulation Solver
Performs numerical integration of the 4-compartment pulmonary PK/PD ODE system
using scipy.integrate.solve_ivp across the 0-24 hour timeline.
"""

import numpy as np
from scipy.integrate import solve_ivp
from typing import Dict, Any

from models.aerosol import calculate_regional_deposition
from models.pharmacokinetics import PKParameters, compute_effective_rates, pk_derivatives
from models.pharmacodynamics import calculate_efficacy, calculate_auc

def run_simulation(
    mmad: float = 2.5,
    inhaled_dose: float = 10.0,
    mucus_viscosity_factor: float = 1.0,
    mucociliary_clearance_factor: float = 1.0,
    airway_radius_factor: float = 1.0,
    tidal_volume_factor: float = 1.0,
    k_release: float = 0.35,
    k_clear_normal: float = 0.25,
    k_clear_drug_normal: float = 0.40,
    k_diff_normal: float = 0.45,
    k_blood: float = 0.15,
    k_tissue_loss: float = 0.08,
    k_systemic_clear: float = 0.30,
    diffusion_sensitivity_alpha: float = 0.7,
    emax: float = 100.0,
    ec50: float = 0.5,
    safety_threshold_db: float = 5.0,
    t_end: float = 24.0,
    time_points: int = 300,
    preset_name: str = "Healthy Adult",
) -> Dict[str, Any]:
    """
    Executes the unified computational pipeline:
      Inputs & Presets -> Aerosol Deposition Model (MMAD + Airway caliber + Tidal volume)
      -> Regional Deposition Fractions (Upper G0-G4, Bronchial G5-G16, Alveolar G17-G23)
      -> Initial Pulmonary Reservoir -> 4-Compartment PK Model (solve_ivp RK45)
      -> Target Tissue Concentration -> Emax/Hill PD Model
      -> Systemic Blood Exposure & Safety Threshold -> Visualizations & CSV Export
    """
    # 1. Aerosol Deposition
    deposition = calculate_regional_deposition(
        mmad=mmad,
        inhaled_dose=inhaled_dose,
        airway_radius_factor=airway_radius_factor,
        tidal_volume_factor=tidal_volume_factor,
    )
    
    # Initial Conditions:
    # Conducting bronchial and deep alveolar deposition form the initial mucosal reservoir Nm(0).
    # Upper airway deposition is rapidly swallowed / cleared out of the respiratory tract.
    initial_nm = deposition["pulmonary_dose"]
    initial_dm = 0.0
    initial_dt = 0.0
    initial_db = 0.0
    y0 = [initial_nm, initial_dm, initial_dt, initial_db]
    
    # 2. Pharmacokinetic Parameters & Disease Modifiers
    pk_params = PKParameters(
        k_release=k_release,
        k_clear_normal=k_clear_normal,
        k_clear_drug_normal=k_clear_drug_normal,
        k_diff_normal=k_diff_normal,
        k_blood=k_blood,
        k_tissue_loss=k_tissue_loss,
        k_systemic_clear=k_systemic_clear,
        mucus_viscosity_factor=mucus_viscosity_factor,
        mucociliary_clearance_factor=mucociliary_clearance_factor,
        diffusion_sensitivity_alpha=diffusion_sensitivity_alpha,
    )
    effective_rates = compute_effective_rates(pk_params)
    
    # 3. ODE Integration using scipy.integrate.solve_ivp (RK45)
    t_eval = np.linspace(0.0, t_end, time_points)
    
    sol = solve_ivp(
        fun=lambda t, y: pk_derivatives(t, y, effective_rates),
        t_span=(0.0, t_end),
        y0=y0,
        t_eval=t_eval,
        method="RK45",
        rtol=1e-6,
        atol=1e-8,
    )
    
    times = sol.t
    nm = np.maximum(0.0, sol.y[0])
    dm = np.maximum(0.0, sol.y[1])
    dt = np.maximum(0.0, sol.y[2])
    db = np.maximum(0.0, sol.y[3])
    
    # 4. Pharmacodynamics Calculation
    efficacy = calculate_efficacy(dt, emax=emax, ec50=ec50)
    
    # 5. Key Pharmacometric & Safety Metrics
    alveolar_deposition_pct = deposition["f_alveolar"] * 100.0
    upper_deposition_pct = deposition["f_upper"] * 100.0
    bronchial_deposition_pct = deposition["f_bronchial"] * 100.0
    
    peak_tissue_conc = float(np.max(dt))
    t_max_tissue_idx = int(np.argmax(dt))
    t_max_tissue = float(times[t_max_tissue_idx])
    
    max_efficacy = float(np.max(efficacy))
    current_efficacy = float(efficacy[-1])
    
    peak_blood_conc = float(np.max(db))
    current_blood_conc = float(db[-1])
    t_max_blood_idx = int(np.argmax(db))
    t_max_blood = float(times[t_max_blood_idx])
    
    mucociliary_loss_pct = float(effective_rates.mucociliary_loss_pct)
    auc_tissue = calculate_auc(times, dt)
    auc_blood = calculate_auc(times, db)
    
    # Time above 50% efficacy threshold
    dt_step = times[1] - times[0] if len(times) > 1 else 0.0
    above_50 = efficacy >= 50.0
    time_above_50 = float(np.sum(above_50) * dt_step) if np.any(above_50) else 0.0
    
    # Time in Therapeutic Window (50% <= E <= 90% AND Db <= safety_threshold_db)
    in_therapeutic_window = (efficacy >= 50.0) & (efficacy <= 90.0) & (db <= safety_threshold_db)
    time_in_therapeutic_window = float(np.sum(in_therapeutic_window) * dt_step) if np.any(in_therapeutic_window) else 0.0
    
    # Safety Hazard Analysis: Db > safety_threshold_db
    above_safety_thresh = db > safety_threshold_db
    is_safety_hazard = bool(np.any(above_safety_thresh))
    time_above_safety_thresh = float(np.sum(above_safety_thresh) * dt_step) if is_safety_hazard else 0.0
    
    metrics = {
        "alveolar_deposition_pct": alveolar_deposition_pct,
        "upper_deposition_pct": upper_deposition_pct,
        "bronchial_deposition_pct": bronchial_deposition_pct,
        "peak_tissue_conc": peak_tissue_conc,
        "t_max_tissue": t_max_tissue,
        "max_efficacy": max_efficacy,
        "current_efficacy": current_efficacy,
        "mucociliary_loss_pct": mucociliary_loss_pct,
        "auc_tissue": auc_tissue,
        "auc_blood": auc_blood,
        "peak_blood_conc": peak_blood_conc,
        "current_blood_conc": current_blood_conc,
        "t_max_blood": t_max_blood,
        "time_above_50": time_above_50,
        "time_in_therapeutic_window": time_in_therapeutic_window,
        "safety_threshold_db": safety_threshold_db,
        "is_safety_hazard": is_safety_hazard,
        "time_above_safety_thresh": time_above_safety_thresh,
        "k_clear_eff": effective_rates.k_clear_eff,
        "k_diff_eff": effective_rates.k_diff_eff,
    }
    
    return {
        "times": times,
        "nm": nm,
        "dm": dm,
        "dt": dt,
        "db": db,
        "efficacy": efficacy,
        "deposition": deposition,
        "effective_rates": effective_rates,
        "metrics": metrics,
        "inputs": {
            "preset_name": preset_name,
            "mmad": mmad,
            "inhaled_dose": inhaled_dose,
            "mucus_viscosity_factor": mucus_viscosity_factor,
            "mucociliary_clearance_factor": mucociliary_clearance_factor,
            "airway_radius_factor": airway_radius_factor,
            "tidal_volume_factor": tidal_volume_factor,
            "k_release": k_release,
            "k_clear_normal": k_clear_normal,
            "k_clear_drug_normal": k_clear_drug_normal,
            "k_diff_normal": k_diff_normal,
            "k_blood": k_blood,
            "k_tissue_loss": k_tissue_loss,
            "k_systemic_clear": k_systemic_clear,
            "diffusion_sensitivity_alpha": diffusion_sensitivity_alpha,
            "emax": emax,
            "ec50": ec50,
            "safety_threshold_db": safety_threshold_db,
            "t_end": t_end,
        },
    }

File: src/simulation/solver.ts
import { SimulationParameters, TimePointResult, SimulationMetrics, RegionalDeposition } from '../types';
import { calculateRegionalDeposition } from '../models/aerosol';
import { computeEffectiveRates, evaluateDerivatives, EffectiveRates } from '../models/pharmacokinetics';
import { calculateEfficacy, calculateAUC } from '../models/pharmacodynamics';

export interface SimulationOutput {
  timeSeries: TimePointResult[];
  metrics: SimulationMetrics;
  deposition: RegionalDeposition;
  effectiveRates: EffectiveRates;
}

/**
 * High-precision 4th-Order Runge-Kutta (RK4) numerical ODE solver
 * designed to emulate scipy.integrate.solve_ivp across the 0-24h (or user defined) domain.
 */
export function runSimulation(params: SimulationParameters): SimulationOutput {
  // 1. Aerosol deposition calculation
  const deposition = calculateRegionalDeposition(params.mmad, params.inhaledDose);

  // Initial condition: Nanoparticles deposited in the respiratory tract
  // Upper airway fraction is swallowed/cleared directly via oropharynx;
  // conducting bronchial + deep alveolar fractions enter mucosal reservoir Nm(0).
  const initialNm = deposition.totalLungDose;
  const initialDm = 0;
  const initialDt = 0;
  const initialDb = 0;

  // 2. Compute modified kinetic rates based on disease factor V and alpha
  const effectiveRates = computeEffectiveRates(params);

  // 3. RK4 integration setup
  const tStart = 0;
  const tEnd = params.simulationHours || 24;
  const numSteps = 240; // 0.1h step for smooth 24h trajectory
  const h = (tEnd - tStart) / numSteps;

  const timeSeries: TimePointResult[] = [];

  let currentState: [number, number, number, number] = [initialNm, initialDm, initialDt, initialDb];
  let currentTime = tStart;

  // Record t = 0
  const initialEfficacy = calculateEfficacy(initialDt, params.emax, params.ec50);
  timeSeries.push({
    time: 0,
    nm: parseFloat(initialNm.toFixed(4)),
    dm: parseFloat(initialDm.toFixed(4)),
    dt: parseFloat(initialDt.toFixed(4)),
    db: parseFloat(initialDb.toFixed(4)),
    efficacy: parseFloat(initialEfficacy.toFixed(2)),
    totalPulmonaryDrug: parseFloat((initialNm + initialDm + initialDt).toFixed(4)),
  });

  for (let i = 0; i < numSteps; i++) {
    // RK4 step
    const k1 = evaluateDerivatives(currentTime, currentState, effectiveRates);

    const stateK2: [number, number, number, number] = [
      Math.max(0, currentState[0] + 0.5 * h * k1[0]),
      Math.max(0, currentState[1] + 0.5 * h * k1[1]),
      Math.max(0, currentState[2] + 0.5 * h * k1[2]),
      Math.max(0, currentState[3] + 0.5 * h * k1[3]),
    ];
    const k2 = evaluateDerivatives(currentTime + 0.5 * h, stateK2, effectiveRates);

    const stateK3: [number, number, number, number] = [
      Math.max(0, currentState[0] + 0.5 * h * k2[0]),
      Math.max(0, currentState[1] + 0.5 * h * k2[1]),
      Math.max(0, currentState[2] + 0.5 * h * k2[2]),
      Math.max(0, currentState[3] + 0.5 * h * k2[3]),
    ];
    const k3 = evaluateDerivatives(currentTime + 0.5 * h, stateK3, effectiveRates);

    const stateK4: [number, number, number, number] = [
      Math.max(0, currentState[0] + h * k3[0]),
      Math.max(0, currentState[1] + h * k3[1]),
      Math.max(0, currentState[2] + h * k3[2]),
      Math.max(0, currentState[3] + h * k3[3]),
    ];
    const k4 = evaluateDerivatives(currentTime + h, stateK4, effectiveRates);

    // Update state with non-negativity constraint
    currentState = [
      Math.max(0, currentState[0] + (h / 6) * (k1[0] + 2 * k2[0] + 2 * k3[0] + k4[0])),
      Math.max(0, currentState[1] + (h / 6) * (k1[1] + 2 * k2[1] + 2 * k3[1] + k4[1])),
      Math.max(0, currentState[2] + (h / 6) * (k1[2] + 2 * k2[2] + 2 * k3[2] + k4[2])),
      Math.max(0, currentState[3] + (h / 6) * (k1[3] + 2 * k2[3] + 2 * k3[3] + k4[3])),
    ];

    currentTime += h;

    const eff = calculateEfficacy(currentState[2], params.emax, params.ec50);

    timeSeries.push({
      time: parseFloat(currentTime.toFixed(2)),
      nm: parseFloat(currentState[0].toFixed(4)),
      dm: parseFloat(currentState[1].toFixed(4)),
      dt: parseFloat(currentState[2].toFixed(4)),
      db: parseFloat(currentState[3].toFixed(4)),
      efficacy: parseFloat(eff.toFixed(2)),
      totalPulmonaryDrug: parseFloat((currentState[0] + currentState[1] + currentState[2]).toFixed(4)),
    });
  }

  // Calculate summary metrics
  let peakTissueConcentration = 0;
  let tMaxTissue = 0;
  let maxPulmonaryEfficacy = 0;
  let cMaxBlood = 0;
  let timeAbove50PctEfficacy = 0;

  for (let i = 0; i < timeSeries.length; i++) {
    const pt = timeSeries[i];
    if (pt.dt > peakTissueConcentration) {
      peakTissueConcentration = pt.dt;
      tMaxTissue = pt.time;
    }
    if (pt.efficacy > maxPulmonaryEfficacy) {
      maxPulmonaryEfficacy = pt.efficacy;
    }
    if (pt.db > cMaxBlood) {
      cMaxBlood = pt.db;
    }
    if (pt.efficacy >= 50) {
      // trapezoidal time accumulation
      timeAbove50PctEfficacy += (i > 0 ? pt.time - timeSeries[i - 1].time : 0);
    }
  }

  const aucTissue = calculateAUC(timeSeries.map((t) => ({ time: t.time, val: t.dt })));
  const aucBlood = calculateAUC(timeSeries.map((t) => ({ time: t.time, val: t.db })));

  const metrics: SimulationMetrics = {
    alveolarDepositionPct: parseFloat((deposition.alveolar * 100).toFixed(1)),
    peakTissueConcentration: parseFloat(peakTissueConcentration.toFixed(3)),
    tMaxTissue: parseFloat(tMaxTissue.toFixed(1)),
    maxPulmonaryEfficacy: parseFloat(maxPulmonaryEfficacy.toFixed(1)),
    mucociliaryLossPct: parseFloat(effectiveRates.mucociliaryLossPct.toFixed(1)),
    aucTissue: parseFloat(aucTissue.toFixed(3)),
    aucBlood: parseFloat(aucBlood.toFixed(3)),
    timeAbove50PctEfficacy: parseFloat(timeAbove50PctEfficacy.toFixed(1)),
    cMaxBlood: parseFloat(cMaxBlood.toFixed(3)),
    effectiveKClear: parseFloat(effectiveRates.kClearEff.toFixed(3)),
    effectiveKDiff: parseFloat(effectiveRates.kDiffEff.toFixed(3)),
  };

  return {
    timeSeries,
    metrics,
    deposition,
    effectiveRates,
  };
}
